kratos/processes/calculate_discontinuous_distance_to_skin_process.cpp

```diff
@@ -27,522 +27,744 @@
+
+    template<std::size_t TDim>
+    void CalculateDiscontinuousDistanceToSkinProcess<TDim>::CalculateElementalDistances(
+        Element& rElement1,
+        PointerVector<GeometricalObject>& rIntersectedObjects)
+    {
+        if (rIntersectedObjects.empty()) {
+            rElement1.Set(TO_SPLIT, false);
+            return;
+        }
+
+        // Get edges container
+        const auto r_edges_container = rElement1.GetGeometry().GenerateEdges();
+
+        // Compute the number of intersected edges
+        array_1d<double, mNumEdges> cut_edges_ratio_vector;
+        array_1d<double, mNumEdges> cut_extra_edges_ratio_vector;
+        std::vector<array_1d <double,3> > int_pts_vector;
+        const unsigned int n_cut_edges = ComputeEdgesIntersections(rElement1, rIntersectedObjects, r_edges_container,
+            cut_edges_ratio_vector, cut_extra_edges_ratio_vector, int_pts_vector);
+
+        // Check if there is intersection: 3 or more intersected edges for a tetrahedron
+        // If there is only 1 or 2 intersected edges, intersection is not considered
+        // If there is intersection, calculate the elemental distances
+        const bool is_intersection = (n_cut_edges < rElement1.GetGeometry().WorkingSpaceDimension()) ? false : true;
+        if (is_intersection){
+            ComputeIntersectionPlaneElementalDistances(rElement1, rIntersectedObjects, int_pts_vector);
+        }
+
+        // Check if the element is split and set the TO_SPLIT flag accordingly
+        const double epsilon = std::numeric_limits<double>::epsilon();
+        SetToSplitFlag(rElement1, epsilon);
+    }
+
+    template<std::size_t TDim>
+    void CalculateDiscontinuousDistanceToSkinProcess<TDim>::CalculateElementalAndEdgeDistances(
+        Element& rElement1,
+        PointerVector<GeometricalObject>& rIntersectedObjects)
+    {
+        if (rIntersectedObjects.empty()) {
+            rElement1.Set(TO_SPLIT, false);
+            return;
+        }
+
+        // Get edges container
+        const auto r_edges_container = rElement1.GetGeometry().GenerateEdges();
+
+        // Compute the number of intersected edges
+        array_1d<double, mNumEdges> cut_edges_ratio_vector;
+        array_1d<double, mNumEdges> cut_extra_edges_ratio_vector;
+        std::vector<array_1d <double,3> > int_pts_vector;
+        const unsigned int n_cut_edges = ComputeEdgesIntersections(rElement1, rIntersectedObjects, r_edges_container,
+            cut_edges_ratio_vector, cut_extra_edges_ratio_vector, int_pts_vector);
+
+        // Save the cut edges ratios in the ELEMENTAL_EDGE_DISTANCES variable
+        rElement1.GetValue(ELEMENTAL_EDGE_DISTANCES) = cut_edges_ratio_vector;
+
+        // Check if there is an intersection
+        bool is_intersection = false;
+        // Extrapolated edge distances were calculated (for Ausas incised elements)
+        if (mOptions.Is(CalculateDiscontinuousDistanceToSkinProcessFlags::CALCULATE_ELEMENTAL_EDGE_DISTANCES_EXTRAPOLATED)) {
```

Comment:
What happens if the user only sets `CALCULATE_ELEMENTAL_EDGE_DISTANCES` (i.e. edge distances without interpolated values)? I think that this breaks the old behavior.

Reply:
Then, the extrapolated edge distances and the elemental distances based on actual and extrapolated edge ratios will not be calculated and we will use the "standard" way to check if there is an intersection.
Probably, that wasn't what you were asking about.. if it's not clear we could do a quick meeting.

Reply:
Let me try to explain myself in a different way. We'd like to have the next three behaviors:
1) Standard case (both flags are off)
2) Edge distances without projected values (`CALCULATE_ELEMENTAL_EDGE_DISTANCES` on)
3) Edge distances with projected values (both `CALCULATE_ELEMENTAL_EDGE_DISTANCES` and `CALCULATE_ELEMENTAL_EDGE_DISTANCES_EXTRAPOLATED` on)
I'm not sure if this can be achieved from the code above.

Reply:
Yes, I agree with the behavior we wish for and I am pretty sure it is achieved with that code, although maybe not in the nicest way.

Reply:
Actually the tests should give an errors if we wouldn't achieve that behavior, otherwise the tests wouldn't be good enough, right? And again, I am pretty sure, that they would give errors, particularly because I did get errors, when it didn't work the way we wanted it too, for example when I was figuring out how to work with two flags in one variable (Flags mOptions).

Reply:
OK. Then perfect. Resolving.